File: leetcode_problems/typescript/src/lc1640_check_array_formation_through_concatenation.ts
function canFormArray(arr: number[], pieces: number[][]): boolean {
  let map: Map<number, number> = new Map();
  for (let i = 0; i < arr.length; i++) {
    map.set(arr[i], i);
  }
  for (let piece of pieces) {
    if (!map.has(piece[0])) {
      return false;
    }
    let prev = map.get(piece[0]);
    for (let j = 1; j < piece.length; j++) {
      if (!map.has(piece[j]) || map.get(piece[j]) !== prev! + 1) {
        return false;
      }
      prev = map.get(piece[j]);
    }
  }
  return true;
}

File: leetcode_problems/cpp/lc1640_check_array_formation_through_concatenation.cpp
#include <vector>
#include <iostream>
#include <unordered_map>
using std::unordered_map;
using std::vector;

class Solution
{
public:
    bool canFormArray(vector<int> &arr, vector<vector<int>> &pieces)
    {
        unordered_map<int, int> map;
        for (int i = 0; i < arr.size(); i++)
        {
            map[arr[i]] = i;
        }
        for (const auto &piece : pieces)
        {
            if (map.find(piece[0]) == map.end())
            {
                return false;
            }
            int prev = map[piece[0]];
            for (int j = 1; j < piece.size(); j++)
            {
                if (map.find(piece[j]) == map.end() || map[piece[j]] != prev + 1)
                {
                    return false;
                }
                prev = map[piece[j]];
            }
        }
        return true;
    }
};
